Lib/saslprep.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2016-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""An implementation of RFC4013 SASLprep."""
+from typing import Any, Optional
+
+try:
+    import stringprep
+except ImportError:
+    HAVE_STRINGPREP = False
+
+    def saslprep(data: Any, prohibit_unassigned_code_points: Optional[bool] = True) -> str:
+        """SASLprep dummy"""
+        if isinstance(data, str):
+            raise TypeError(
+                "The stringprep module is not available. Usernames and "
+                "passwords must be instances of bytes."
+            )
+        return data
+
+else:
+    HAVE_STRINGPREP = True
+    import unicodedata
+
+    # RFC4013 section 2.3 prohibited output.
+    _PROHIBITED = (
+        # A strict reading of RFC 4013 requires table c12 here, but
+        # characters from it are mapped to SPACE in the Map step. Can
+        # normalization reintroduce them somehow?
+        stringprep.in_table_c12,
+        stringprep.in_table_c21_c22,
+        stringprep.in_table_c3,
+        stringprep.in_table_c4,
+        stringprep.in_table_c5,
+        stringprep.in_table_c6,
+        stringprep.in_table_c7,
+        stringprep.in_table_c8,
+        stringprep.in_table_c9,
+    )
+
+    def saslprep(data: Any, prohibit_unassigned_code_points: Optional[bool] = True) -> str:
+        """An implementation of RFC4013 SASLprep.
+
+        :Parameters:
+          - `data`: The string to SASLprep. Unicode strings
+            (:class:`str`) are supported. Byte strings
+            (:class:`bytes`) are ignored.
+          - `prohibit_unassigned_code_points`: True / False. RFC 3454
+            and RFCs for various SASL mechanisms distinguish between
+            `queries` (unassigned code points allowed) and
+            `stored strings` (unassigned code points prohibited). Defaults
+            to ``True`` (unassigned code points are prohibited).
+
+        :Returns:
+        The SASLprep'ed version of `data`.
+        """
+        prohibited: Any
```

Comment:
This does nothing.
```suggestion
```

Reply:
This seems to be there to silence a type checker.

Type checkers in Python consider tuples structures, not immutable lists. Therefore concatenating tuples changes their type.

Instead of removing the line, I'd provide a better annotation.

Something like:
```suggestion
        prohibited: tuple[Callable[[str], bool], ...]
```